Sources/ProjectDescription/Template/AttributeHelpers.swift

```diff
@@ -0,0 +1,36 @@
+import Foundation
+
+/// Error that occurs when parsing attributes from input
+enum ParsingError: Error, CustomStringConvertible, Equatable {
+    /// Thrown when could not find attribute in input
+    case attributeNotFound(String)
+    /// Thrown when attributes not provided
+    case attributesNotProvided
+
+    var description: String {
+        switch self {
+        case let .attributeNotFound(name):
+            return "Could not find attribute \(name)"
+        case .attributesNotProvided:
+            return "You must provide --attributes option"
+        }
+    }
+}
+
+/// Returns value for `ParsedAttribute` of `name` from user input
+/// - Parameters:
+///     - name: Name of `ParsedAttribute`
+///     - arguments: Arguments to get attribute from (usually you should lead this at default)
+/// - Returns: Value of `ParsedAttribute`
+public func getAttribute(for name: String, arguments: [String] = CommandLine.arguments) throws -> String {
```

Comment:
Models in the `ProjectDescription` framework are intentionally designed to be dumb: they contain a constructor, attributes, and the logic to encode and decode them. All the business logic lives on the `TuistLoader` and the `TuistGenerator` side.

This implementation is not consistent with that. Models parse information passed as shell arguments, that they are then using to substitute variables in the template. I'd suggest changing the approach.

What Tuist gets from the `ProjectDescription` is just a description of the template, being the template metadata with references to files or content that might contain attributes. To be able to substitute those variables, we need them to have a special format within the template. That's something up to us to define, but it could be something along the lines of:

```swift
// Swift file

class {{name:string:true}}Interactor {

}
```

Or easier, we can follow [Stencil's template system](https://github.com/stencilproject/Stencil). After loading the template in memory, we can parse the arguments and do the substitution of the variables. If there are missing variables or mismatching of types, we can fail and exit with an error. If the substitution runs successfully, then we pass the template down to the scaffolding utility to generate the template.

What do you think @fortmarek ?

Reply:
Thanks for the thorough comment @pepibumur 🙂 

This bit of logic was needed in order to be able to define custom templates in Swift - which has its benefits like not forcing developers to learn a template language (though I realize that Stencil is a de facto standard in this regard) and as in all other Tuist manifests the ability to leverage Swift as a language with variables and type-safety. Besides that there is for example a possibility for the user to define custom errors right in the template - shared with `ProjectDescriptionHelpers` across the templates.

That being said, I agree that keeping `ProjectDescription` simple as possible is definitely a high priority. 

If you think that trying to work around it is worth the effort, I can have another look if there is a way how to get rid of the business logic here (or at least to the point where it is simple as, say, a decoding). Otherwise, we can definitely fall back to Stencil, I have tried to take a different approach, but I am not at all saying it is definitely the right one.

Reply:
Hey @fortmarek 

> This bit of logic was needed in order to be able to define custom templates in Swift - which has its benefits like not forcing developers to learn a template language (though I realize that Stencil is a de facto standard in this regard) and as in all other Tuist manifests the ability to leverage Swift as a language with variables and type-safety.

I see where your rationale came from. I think it's fine adopting a template system like Stencil. It's very standard and easy to learn. The benefit of using it is not just simplifying the models in the `ProjectDescription` framework, but also being able to use the same template system for both, templates defined in a file or in the `Template.swift` file.

> That being said, I agree that keeping ProjectDescription simple as possible is definitely a high priority.

Not a priority, but I think is good to stay consistent with how we've been doing things. If at some point it proves not to be the best decision going forward, we can reconsider it. It's been working well so far, and I think this feature can be well implemented following the same approach.

> If you think that trying to work around it is worth the effort, I can have another look if there is a way how to get rid of the business logic here (or at least to the point where it is simple as, say, a decoding). Otherwise, we can definitely fall back to Stencil, I have tried to take a different approach, but I am not at all saying it is definitely the right one.

I'd suggest towards Stencil and doing the replacement of the variables after the template has been loaded.